### PR TITLE
REF: use maybe_convert_objects in pd.array

### LINKED ISSUE #29973: REF: Deduplicate array, infer_dtype, and maybe_convert_objects
There's some duplicate / slightly diverging functionality between `pd.array`, `infer_dtype` and `convert_objects`. Both `pd.array` and `convert_objects` are roughly "convert this list-like thing to this better array". See the conversation in  https://github.com/pandas-dev/pandas/pull/29799#discussion_r352255141 for more.

Hopefully this can be deduplicated.

cc @jreback.

Followup to https://github.com/pandas-dev/pandas/pull/29799

### LINKED ISSUE #52972: PERF: use maybe_convert_objects instead of infer_dtype in pd.array
I'm not sure if we can get down to single-pass, but I think we can get down to one-fewer.

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #52972 (Replace xxxx with the GitHub issue number)
- [x] closes #29973
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

@phofl the only other place where the convert_nullable_whatever option is used is in the sql code that i think you implemented.  will any of the changes here affect that?

### CODE REVIEW

**Comment by WillAyd** [comment]
@lithomas1 looked at this function recently so might have feedback

**Comment by phofl** [comment]
should be fine for the sql use case

File: pandas/_libs/lib.pyx [hunk 1]
```diff
@@ -2834,6 +2843,22 @@ def maybe_convert_objects(ndarray[object] objects,
...
+
+            # Set these values to 1 to be deterministic, match
+            #  IntegerArray._internal_fill_value
+            result[mask] = 1
+            result = IntegerArray(result, mask)
+        elif result is floats and convert_to_nullable_dtype:
+            from pandas.core.arrays import FloatingArray
+
+            # Set these values to 1.0 to be deterministic, match
+            #  IntegerArray._internal_fill_value
```
Comment (mroeschke) [inline pandas/_libs/lib.pyx:2858]:
```suggestion
            #  FloatingArray._internal_fill_value
```
?

**Comment by jbrockmendel** [comment]
rebased

**Review by mroeschke — APPROVED** [review]

**Pull request merged**